h2o-py/h2o/explanation/_explain.py

```diff
@@ -2650,6 +2650,163 @@ def learning_curve_plot(
+    Pareto front there isn't a model that would be better in both criteria. This can be useful for example in picking
+    models that are fast to predict and at the same time have high accuracy. For generic data.frames/H2OFrames input
+    the task is assumed to be minimization for both metrics.
+
+    :param models: H2OAutoML, H2OGrid, list of models, or leaderboard frame
+    :param x_metric: metric present in the leaderboard
+    :param y_metric: metric present in the leaderboard
+    :param title: title used for the plot
+    :param color_col: Categorical column in the leaderboard that should be used for coloring the points
+    :param figsize: figure size; passed directly to matplotlib
+    :param colormap: colormap to use
+    :return: object that contains the resulting figure (can be accessed using result.figure())
+
+    :examples:
+    >>> import h2o
+    >>> from h2o.automl import H2OAutoML
+    >>>
+    >>> h2o.init()
+    >>>
+    >>> # Import the wine dataset into H2O:
+    >>> f = "https://h2o-public-test-data.s3.amazonaws.com/smalldata/wine/winequality-redwhite-no-BOM.csv"
+    >>> df = h2o.import_file(f)
+    >>>
+    >>> # Set the response
+    >>> response = "quality"
+    >>>
+    >>> # Split the dataset into a train and test set:
+    >>> train, test = df.split_frame([0.8])
+    >>>
+    >>> # Train an H2OAutoML
+    >>> aml = H2OAutoML(max_models=10)
+    >>> aml.train(y=response, training_frame=train)
+    >>>
+    >>> # Create the Pareto front
+    >>> pf = aml.pareto_front()
+    >>> pf.figure() # get the Pareto front plot
+    >>> pf # H2OFrame containing the Pareto front subset of the leaderboard
+    """
+    plt = get_matplotlib_pyplot(False, True)
+    from matplotlib.lines import Line2D
+    name = None
+    if isinstance(models, h2o.automl._base.H2OAutoMLBaseMixin):
+        leaderboard = _get_leaderboard(models, None, top_n=float("inf"))
+        name = models.project_name
+    elif isinstance(models, h2o.grid.grid_search.H2OGridSearch):
+        leaderboard = _get_leaderboard(models.models, None, top_n=float("inf"))
+        name = models.grid_id
+    elif isinstance(models, list):
+        models = [h2o.get_model(m) if is_type(m, str) else m for m in models]
+        leaderboard = _get_leaderboard(models, None, top_n=float("inf"))
+    elif isinstance(models, h2o.H2OFrame):
+        leaderboard = models
+    else:
+        try:
+            leaderboard = h2o.H2OFrame(models)
+        except Exception:
+            raise ValueError("`models` paraameter has to be either H2OAutoML, H2OGrid, list of models or coercible to H2OFrame!")
+
+    if x_metric is None:
+        if "predict_time_per_row_ms" in leaderboard.names:
+            x_metric = "predict_time_per_row_ms"
+        else:
+            x_metric = leaderboard.names[2] if "model_id" in leaderboard.names else leaderboard.names[0]
+    if y_metric is None:
+        y_metric = leaderboard.names[1]
```

Comment:
It's not just AutoML vs list of models but also a "leaderboard" (some frame with metrics that can be used to pick the leader) with some additional metrics, e.g., leaderboard + fairness metrics. Or even just a data frame with leader models from different AutoMLs  etc.

In other words, the defaults are complicated because we allow passing a data frame and the data frame can have  the prediction duration present or not. This is likely if the user extends our extended leaderboard with their own metrics.

Reply:
> the defaults are complicated because ...

the defaults are complicated because you change the semantic of your function depending on the first parameter.

I think it's perfectly ok to overload params in Py with the condition that it's only used as syntactic sugar. If your function starts to behave differently, then there's something wrong.
In this case:
```python
aml.pareto_front()
pareto_front(aml)
# vs
pareto_front(aml.leaderboard)
```
will produce 2 different plots.
I don't think this should be the case:
1. counter-intuitive.
2. forces `pareto_front` to have too much knowledge about optional AutoML features (way outside its responsibility).
3. locks the default AutoML pareto front to the need of the day: what if in 3 months, we're mainly interested in accuracy vs. fairness? or something else?

If the AutoML defaults are not "natural", then it's AutoML's responsibility to provide that logic:
```python
pareto_front(aml.accuracy_vs_complexity)
pareto_front(aml.accuracy_vs_fairness)

aml.pareto_front()  # not sure it even makes sense but if it does, it should delegate aml specificities to aml itself, e.g.:
def pareto_front(obj, **kwargs):
   if hasattr(obj, 'pareto_front_frame'):
      pff = obj.pareto_front_frame(**kwargs)  # also provides user with a data allowing to easily produce its own PF plot.
   else:
      pff = ... # standard dumb defaults
   # pff should only contain data necessary for the PF plot
   # now rendering logic only
```
I think that leaking automl or leaderboard logic all over the place can only create inconsistencies and issues in the future.

Reply:
I'm not really sure how to approach this. Would throwing an error "You need to specify x_metric, y_metric" be better in the `pareto_front(aml.leaderboard)` case? 
It still changes the behavior of the function depending on the param but if we want the defaults to use the prediction speed the `aml.pareto_front()` and `pareto_front(aml.leaderboard)` have to differ as the latter doesn't have the necessary information.

Reply:
I just feel like this function has too many responsibilities, trying to solve all the use cases in one place, which for me leads to confusion.
I think the only fundamental params for `pareto_front` should be:
1. a frame with at least 3 columns or `None`: if None then the 3 following params are mandatory and should all be lists.
2. labels: None, or string (frame column name), or list/single col (of string). if None, then use frame first col.  If string, then frame should be provided. If list, then frame should be None. Otherwise error
3. x_metric: None, or string, or list/single col (same as above). If None, then use frame second col, otherwise same as above.
4. y_metric: None, or string, or list/single col (same as above). If None, then use frame third col.

With this, you can now apply your method to anything without trying to be too clever using the inversion of control suggested above (`if hasattr(obj, 'pareto_front_frame')`): aml can provide its default behavior, grid can provide its default and so on... and they can also easily extend the semantic thanks to the `**kwargs` without polluting PF code:
```python
aml.pareto_front(type='accuracy_vs_complexity')  # default
aml.pareto_front(type='accuracy_vs_fairness')
...
grid.pareto_front() # if we don't have good defaults, then fail asking for specific metrics to compare
```

Regarding `pareto_front(aml.leaderboard)`, as soon as we don't document the function by saying that it accepts leaderboards, then I don't think it should have any specific behaviour: it just behaves like a frame as suggested above.
Fundamentally, `pareto_front(...)` should be documented to describe the behaviour with frames.
And when assigning things like
```python
h2o.automl._base.H2OAutoMLBaseMixin.pareto_front = pareto_front
```
you could instead duplicate the fn and add a more specific doc, for example:
```python
h2o.automl._base.H2OAutoMLBaseMixin.pareto_front = update_wrapper(partial(pareto_front), pareto_front)
h2o.automl._base.H2OAutoMLBaseMixin.pareto_front.__doc__ = "describe behaviour when used on AutoML instance"
```
(btw, imo, the decoration above should be done in automl module, not in explain: automl feature is the only one deciding if it needs this feature or not).
It keeps `pareto_front` function cleaner, with simple expectations, and delegate specific behaviour to those objects that need it.
wdyt?

Reply:
or just simply, no callback in `explain.pareto_front` but directly add a `pareto_front()` method to automl and grids that define its own defaults instead of all this glue.

Reply:
Since we now have `h2o.make_leaderboard`, I am thinking about following behavior:

**In Python:**

- `h2o.explanation.pareto_front` - takes just a frame, metrics, location of the optimum and title for the plot. This function is intended to be used with any frame including the output of `h2o.make_leaderboard`

- `aml.pareto_front` - has an optional `leaderboard_frame` if that's not specified it will use leaderboard obtained by `aml.get_leaderboard("ALL")` otherwise it will create a new leaderboard using the `leaderboard_frame`. This is then passed to the `h2o.explanation.pareto_front`.

- `grid.pareto_front`  - requires the use of `leaderboard_frame` when the `predict_time_per_row_ms` is used otherwise when the metrics are specified (and don't need the `leaderboard_frame` for their calculation) it does also work without the `leaderboard_frame`.

This way I can have "nice" defaults for AutoML and Grid (since both will always have `predict_time_per_row_ms` available due to the requirement of the `leaderboard_frame`) and a general `pareto_front` function that deals just with frames. 

**In R:**

- `h2o.pareto_front` - accepts AutoML, Grid or frame. When used with AutoML and Grid a user can specify also the `leaderboard_frame`.  Basically it behaves the same way  as python's version just all in one function.

I feel like this is relatively common in R but I might be wrong. Should I try to reimplement similar behavior as in python using S3 or S4?

---

Or we can remove the defaults and have just the simple `h2o.explanation.pareto_front` that would accept just the "leaderboard"(or any other frame), x,y metrics, location of the optimum, title and other things used just for plotting. I would prefer the simplicity and "explicitness" but it wouldn't be as user friendly:
```python
aml.pareto_front()
# vs 
h2o.explanation.pareto_front(aml.get_leaderboard("ALL"), "predict_time_per_row_ms", "auc")
```
```python
aml.pareto_front(test_frame)
# vs 
h2o.explanation.pareto_front(h2o.make_leaderboard(aml, test_frame, extra_columns="ALL"), "predict_time_per_row_ms", "auc")
```
but the more complicated scenarios would be the same: 
```python
h2o.explanation.pareto_front(h2o.make_leaderboard([aml, grid, gbm_model], 
                                                  test_frame, extra_columns="ALL"), 
                             "predict_time_per_row_ms", "auc")
```

WDYT @ledell @sebhrusen ?